### PR TITLE
⚡️ Enable training with CUDA

### PR DESCRIPTION
**Pull request by emprice**
PR implementing the feature suggested in #58. The integer matmul operation needed for a masked autoregressive transform should be explicitly carried out on CPU and then migrated to the current default device. All other tensors get created on the default device anyway. This way, if the user calls `torch.set_default_device('cuda')`, that preference will be followed.

I have modified existing tests to make sure that no existing code should be broken by my modification; all tests are now performed on CPU and with CUDA to be sure they pass either way.

### CODE REVIEW

**Comment by emprice** [comment]
Looks like `torch.set_default_device()` and `torch.get_default_device()` are newer than I realized. `torch.set_default_device()` is only used for testing, and those could easily be conditionally skipped. Not sure how to fix the issue with `torch.get_default_device()`. Converting to a NumPy array and then using `torch.tensor` is a possibility as a workaround.

File: zuko/nn.py [hunk 1]
```diff
@@ -270,7 +270,14 @@ def __init__(
...
         # P_ij = 1 if A_ik = 1 for all k such that A_jk = 1
-        precedence = adjacency.int() @ adjacency.int().t() == adjacency.sum(dim=-1)
+        # PyTorch doesn't support this operation for integer arrays on CUDA devices
+        precedence = (
+            adjacency.cpu().int() @ adjacency.cpu().int().t() == adjacency.sum(dim=-1).cpu()
+        )
+        try:
+            precedence = precedence.to(torch.get_default_device())
+        except AttributeError:
+            precedence = torch.tensor(precedence.detach().cpu().numpy())
```
Comment (francois-rozet) [inline zuko/nn.py:280]: The cast to `.int()` is here because matmul is not possible for boolean tensors on CPU (and CUDA). We can simply cast to `.double()` instead to support CUDA.

**Review by francois-rozet — COMMENTED** [review]
Thank you for your PR. Unfortunately, I think there is a better approach to address #58.

The cast to `.int()` in `MaskedMLP` is there because matmul is not possible for boolean tensors on CPU (and CUDA). We can simply cast to `.double()` instead to support CUDA. I tried in https://github.com/probabilists/zuko/commit/b5fc6cc47d427db994bf38657b641012ab1fd22d and it works perfectly.

Concerning the CI, instead of modifying many (but not all) tests, it would be better to add a global option to `pytest` to run all tests with `set_device_default("cuda")`. I tried in https://github.com/probabilists/zuko/commit/2f325e4c7691ee4f850eaecb8dbdb24ea97216cd, which allows to run

```
pytest --device cuda
```

**Comment by emprice** [comment]
No longer needed -- #58 has been fixed!
